### PR TITLE
fix($compile): still trigger `$onChanges` even if the inner value alr…

### PR DESCRIPTION
**Pull request by petebacondarwin**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

fix

**What is the current behavior? (You can also link to an open issue here)**

the `$onChanges` hook is not getting called if the value on the controller is already set to the new value

A2 does trigger this value - see http://plnkr.co/edit/V15jwFA7e5zsWgLVLDKf?p=preview

**What is the new behavior (if this is a feature change)?**

now the hook is triggered

**Does this PR introduce a breaking change?**

no

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

…eady matches the new value

### CODE REVIEW

File: src/ng/compile.js [hunk 1]
```diff
@@ -3233,7 +3237,7 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
             onChangesQueue.push(triggerOnChangesHook);
           }
           // If the has been a change on this property already then we need to reuse the previous value
-          if (changes[key]) {
+          if (isDefined(changes[key])) {
```
Comment (gkalpak) [inline src/ng/compile.js]: Is this necessary ?

File: test/ng/compileSpec.js [hunk 2]
```diff
@@ -3742,6 +3742,34 @@ describe('$compile', function() {
...
+
+        angular.module('my', [])
+          .component('c1', {
+            controller: TestController,
+            bindings: { 'prop1': '<' }
+          });
+
+        module('my');
+        inject(function($compile, $rootScope) {
+          $rootScope.getVal = function() { return $rootScope.val; };
```
Comment (gkalpak) [inline test/ng/compileSpec.js]: This is not used.

File: test/ng/compileSpec.js [hunk 3]
```diff
@@ -3742,6 +3742,34 @@ describe('$compile', function() {
...
+          element = $compile('<c1 prop1="val"></c1>')($rootScope);
+          var isolateScope = element.isolateScope();
+
+          log = [];
```
Comment (gkalpak) [inline test/ng/compileSpec.js]: Nitpick: Not necessary :grin:

File: test/ng/compileSpec.js [hunk 4]
```diff
@@ -3742,6 +3742,34 @@ describe('$compile', function() {
...
+          $rootScope.$apply('val = 1');
+          expect(log.pop()).toEqual({prop1: jasmine.objectContaining({previousValue: undefined, currentValue: 1})});
+
+          isolateScope.$ctrl.prop1 = 2;
```
Comment (gkalpak) [inline test/ng/compileSpec.js]: Nitpick: `element.controller('c1').prop1 = 2` saves one line of code (Yay :disappointed_relieved:)

File: src/ng/compile.js (hunk 1 above)
Comment (petebacondarwin) [inline src/ng/compile.js]: It seemed safer

Comment (petebacondarwin) [inline src/ng/compile.js]: But I guess since the property value must be `undefined` or an `object` it is not strictly necessary

Comment (petebacondarwin) [inline src/ng/compile.js]: Feel free to do a refactor commit to remove it :-)

**Pull request closed without merging**